### DISCUSSION TITLE
How to implement ReAct like agent using promptflow?

### DISCUSSION THREAD

**Discussion by chaliy** [microsoft/promptflow#3485] [Q&A]
An agent that is in a loop will call LLM, to call a function to get extra context, and when enough context, generate an answer. Ideally, the agent should support other common chat features, like chat history, citation and guardrails.

ReAct Paper - https://arxiv.org/abs/2210.03629

**Comment by brynn-code** [comment]
Hi @chaliy, thanks for reaching us. Would you like to write a dummy script to elaborate more about your case and let us know the mainly blocker you met when achieving this?

**Reply by chaliy** [reply]
The main blocker is that it is not clear how to represent loops in promptflow.

```py
tools = [
   get_weather(city: str)
]
def generate_answer_react(input):
 
  context = ""
  
  for iter in range(1, 4):
     result = llm(f"try to answer users question {input}, or call a tool if you need more data. Use context {context}")
     if "answer" in result: 
        return result["answer"]
     elif "tool_call" in result:
         tool_results = call_tool(result["tool_call"])
         context += tool_results
 
input = "what is the weather in Kyiv and Lviv"
generate_answer_react(input)
```

And it will execute something like:

Iter 1: tool_call get_weather("Kyiv")
Iter 2: tool_call get_weather("Lviv") 
Iter 3: return answer

**Reply by brynn-code** [reply]
I see, currently promptflow does not support loop as an operator, but you could leverage the python for-loop in script tools to achieve your goal. A python tool is a python function with `@tool` decorated, so anything works with python should work normally as a tool.

In the case of your example, the tool script shall be something like:

```python
def get_weather(city: str):
   # do some external calls to get the weather of city
   return weather

@tool 
def generate_answer_react(input: str, connection: AzureOpenAIConnection):
  context = ""
  
  for iter in range(1, 4):
     result = llm(f"try to answer users question {input}, or call a tool if you need more data. Use context {context}")
     if "answer" in result: 
        return result["answer"]
     elif "tool_call" in result:
         tool_results = get_weather(....)
         context += tool_results
```

Here is another example flow to classify url type with some python tools in to illustrate how python tool works: https://github.com/microsoft/promptflow/tree/main/examples/flows/standard/web-classification
